Replace CLI to Std Map Plus Individual Memory Shared

RadWolfie opened the pull request:
What's implemented in this pull request:
- Replace cli string to std map, unordered, pairs (much easier to convert then find what we need quicker)
- Enable individual memory shared settings base on `high_resolution_clock::now()`  value as SID.
- Prevent active emulations to run if using same data directory path.

The codebase is currently functional. However, needs investigation for may only allow one emulation at a time to avoid performance issue and false positive reports.

**Feedback are welcome!**

---

CLI arguments:
- exec (arg0)
  - reserved internally (custom `/exec` will be ignored)
- arg1 (arg1)
  - reserved internally (custom `/arg1` will be ignored)
- `/load`
  - absolute xbe file path
- `/hwnd`
  - attach to GUI's window handle for visual output.
- `/dm`
  - debug mode (integer)
- `/df`
  - debug log file path
- `/sid` (Session ID, should not be tampered with)
  - reserved internally

PatrickvL commented:
The goal of the loader is to allocate NO memory, so that reserving addresses ranges will have the highest chance of success.

This code breaks that goal.

RadWolfie commented:
> The goal of the loader is to allocate NO memory, so that reserving addresses ranges will have the highest chance of success.

Right, except the loader project itself is not touched. Only the GUI and emulation project are being touched for memory allocation. (HINT: Data file path is only taking 4 bytes of shared memory.) I don't see any source files change relative to the loader project base on loader's CMakeLists.txt file.


Anyhow, I forgot to add one field of "sizeof" variable to EmuShared.

ergo720 commented on src/common/util/CxbxUtil.cpp:
```diff
@@ -269,3 +269,15 @@ std::string StripSpaces(const std::string& str)
+	else {
+		return "";
+	}
+}
```
This function is identical to `StripSpaces` of above bar the character string searched by `find_first_not_of`. So, instead of creating this new function, you can modify the existing one by making it accept the character string being searched for and renaming it (perhaps `StripChars`?).
```
std::string StripChars(const std::string &str, const char* char_str)
{
	const size_t s = str.find_first_not_of(char_str);
	if (str.npos != s) {
		return str.substr(s, str.find_last_not_of(char_str) - s + 1);
	}
	else {
		return "";
	}
}
```

RadWolfie replied:
Thanks for the remark! 😃 Indeed is much smarter design. It has been fixed in d5a70c0 commit.

RadWolfie commented:
I had corrected `GetProcessID`, renamed to `GetSessionID` to re-use given SID from CLI. Originally affected standalone emulation since they don't provide hwnd.

Plus added a pop up error message if active data directory is in used or not.

RadWolfie commented:
Ready for final review! 😃

PatrickvL commented:
unordered_map_strings is declared twice in the new .CPP files, so could be moved to a header file

PatrickvL commented:
The underlying type for unordered_map_strings appears also independently in the new .cpp files, so these appearances could be replaced by the new typedef.

PatrickvL commented on src/core/kernel/init/CxbxKrnl.cpp:
```diff
@@ -748,8 +723,10 @@ bool HandleFirstLaunch()
-void CxbxKrnlMain(int argc, char* argv[], uint32_t blocks_reserved[384])
+void CxbxKrnlEmulate(uint32_t blocks_reserved[384])
 {
+	std::string tempValue;
```
Limit scope, move declaration down to right before use

RadWolfie replied:
see comment above

PatrickvL commented on src/common/util/cliConverter.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    for (i; i != map_pairs.end();) {
+        // If argument 1 was input, skip it.
+        if (!i->first.compare("arg1")) {
```
What's "arg1" doing here? Please explain in a comment

RadWolfie replied:
Both exec and arg1 keys are reserved. "exec" is arg0 which doesn't need the key part. "arg1" will always not be part of parse to command line string. Since arg1 become useless when start using "load" key.

PatrickvL commented on src/common/util/cliConverter.cpp:
```diff
@@ -0,0 +1,145 @@
+            continue;
+        }
+        // Otherwise, let's mark as invalid input
+        return unordered_map_strings();
```
This is a quick and dirty form of 'error handling' (there's more of those above) - how can the caller distinguish between those faulty scenarios? Would it be possible to somehow log what's wrong with the input?

RadWolfie replied:
It's to prevent abuse from user inputs. I rather keep generic files untouched from our extended log system. There is an error message which will show if converter's map return as empty. Search for `cli_config::GenConfig` both emulator + old GUI has the error message.

RadWolfie commented on src/common/Settings.cpp:
```diff
@@ -203,41 +205,12 @@ bool Settings::Init()
-				return false;
-			}
+		std::string setupFile;
+		m_gui.DataStorageToggle = SetupFile(setupFile);
```
Are you aware `LoadFile` caller will choose either default (which is above and below remarks you had made) or override by settings file value?

When the setting's file doesn't have the value, then default is always be 0. Hence a bug for portable mode.

LukeUsher replied:
References exist specifically for this purpose: If you only want to pass values to a function, you would use a const reference instead.

RadWolfie commented on src/common/util/cliConfig.cpp:
```diff
@@ -0,0 +1,170 @@
+
+    // Check if previous session ID had been set then use it.
+    if (!GetValue(cli_config::sid, &sessionID)) {
+        using namespace std::chrono;
```
Hmm... true, my original concern was text width is over the limit. Plus I was rambling which chrono's sub-classes to use from.

RadWolfie replied:
resolved

RadWolfie commented on src/gui/WndMain.cpp:
```diff
@@ -2261,8 +2269,18 @@ void WndMain::StartEmulation(HWND hwndParent, DebuggerState LocalDebuggerState /
+            cli_config::SetValue(cli_config::debug_mode, g_Settings->m_core.KrnlDebugMode);
+            cli_config::SetValue(cli_config::debug_file, g_Settings->m_core.szKrnlDebug);
+        }
+        cli_config::GenCMD(szProcArgsBuffer);
```
I don't understand what you're saying.

RadWolfie replied:
resolved

RadWolfie commented on src/common/util/cliConverter.cpp:
```diff
@@ -0,0 +1,145 @@
+        return to_string;
+    }
+
+    to_string += "\"" + i->second + "\" ";
```
Our input strings should never have quotes. If they do, then they need to be stripped. Though, I don't see a valid reason to have string with quotes from our codebase.

RadWolfie commented on src/gui/WinMain.cpp:
```diff
@@ -45,6 +47,7 @@ processorArchitecture = '*' publicKeyToken = '6595b64144ccf1df' language = '*'\"
 int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
 {
 	hActiveModule = hInstance; // == GetModuleHandle(NULL); // Points to GUI (Cxbx.exe) ImageBase
+	std::string tempValue;
```
tempValue simply explain itself, aka temporary value. Which can be reuse at any time, that's why it's at the top than relocate tempValue whenever want to use tempValue to get a value or generate string at earlier stage.

RadWolfie commented:
> unordered_map_strings is declared twice in the new .CPP files, so could be moved to a header file

That's only a lazy and shorter width method, header file shouldn't have it anyway.

RadWolfie commented on src/common/util/cliConverter.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    // If map pairs are empty, return empty string.
+    // Or is not exact match for first iterator, then return empty string.
+    if (i == map_pairs.end() || i->first.compare("exec")) {
```
I'll replace it to `// Or "exec" is not in first iterator by requirement, then return empty string.`

RadWolfie replied:
resolved

RadWolfie commented on src/common/Settings.cpp:
```diff
@@ -144,7 +144,9 @@ static struct {
 {
-	return g_exec_filepath.substr(0, g_exec_filepath.find_last_of("\\/"));
+	std::string exec_path;
+	cli_config::GetValue(cli_config::exec, &exec_path);
```
resolved

RadWolfie commented:
Squashed all worthy remarks and ready for merge.

FYI: I have two other branches ready for PR.

PatrickvL commented:
My remaining remarks aren't all that important - merging...
